elastic/apm-server
# sampling: introduce policies

## Motivation/summary

Revise tail-sampling configuration, removing the
default sampling rate and introducing configurable
sampling policies.

Each policy defines criteria to match root transactions,
and a sampling rate. Policies are evaluated in the
order specified, comparing the specified attributes
against incoming root transactions. In the future policies
could also match non-root transactions or even spans,
at some additional cost.

Attributes that can be used for matching are:
 - service.name
 - service.environment
 - trace.name (root transaction name)
 - trace.outcome (root transaction outcome)

Policies may omit any of the attributes. With the
exception of the service name, omitting an attribute
has the effect of grouping all matching traces for
the purposes of tail-sampling. For example, if only
service name is specified, then all traces originating
at that service will be sampled uniformly.

If service name is not specified, then that policy may
match any service. There is a currently hard-coded
limit of 1000 service names that may be matched in
this way. If the limit is reached, traces from any additional
services will not be sampled.

If a root transaction matches none of the policies, then
it will be dropped and the trace will not be sampled.

## Checklist


I have considered changes for:
~- [ ] documentation~ (later)
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~ (will follow up with monitoring, e.g. number of active sampling groups)
~- [ ] telemetry~ (will follow up with reporting whether it's enabled, and number of policies)

## How to test these changes

1. Run apm-server with the following config:

```yaml
apm-server:
  sampling:
    tail:
      enabled: true
      interval: 10s
      policies:
      - trace.name: many
        sample_rate: 0.5
      - trace.name: few
        sample_rate: 0.1
```

2. Send 1000 transactions with the name "many", and 1000 with the name "few".
3. Wait ~10 seconds
4. Check that there are 500 "many" transactions, and 100 "few" transactions, indexed into `apm-*-transaction`:

```
GET /apm-8.0.0-transaction/_search
{
  "size": 0,
  "aggs": {
    "by_name": {
      "terms": {
        "field": "transaction.name",
        "size": 100
      }
    }
  }
}

{
  "took" : 0,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 600,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "by_name" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [
        {
          "key" : "many",
          "doc_count" : 500
        },
        {
          "key" : "few",
          "doc_count" : 100
        }
      ]
    }
  }
}
```

## Related issues

Part of #4185
